### PR TITLE
Trade UserCase

### PR DESCRIPTION
**Pull request by no7thing**

### CODE REVIEW

**Review by Maxlyu254 — APPROVED** [review]
Good job on the setters and getters and the access management for the variables. However, you still need to change your sequence of calls and dependencies. Most importantly, specify where in the classes are you going to alternate the view models.
PS. The term is "use case" instead of "user case"

**Review by adelinewang3 — APPROVED** [review]
The format looks good. One thing to consider is if the get out of jail card is tradable, if it is you might want to implement that part into your code. 
Another rule to consider if you want it implemented is that when trading in monopoly, houses/hotels can't be traded so if a player wants to trade a property, they would have to sell the houses/hotels first before being able to trade.

**Review by sinemsoylemezoglu — DISMISSED** [review]

**Review by Maxlyu254 — APPROVED** [review]
This does not follow clean architecture, but compiles, so pull request approved. Consider rereading notes about clean architecture and use case structure in this diagram: 
https://docs.google.com/drawings/d/1guxgc3iXlnha_ts0hO3Yk_yEdKgGnaPNtL6vQpJy02k/edit?usp=sharing

File: src/main/java/Trade/TradeOutputBoundary.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package Trade;
+import java.io.IOException;
+
+public interface TradeOutputBoundary {
```
Comment (celinayueh) [inline src/main/java/Trade/TradeOutputBoundary.java:4]: Should use the create() method instead.

File: src/main/java/Trade/TradePresenter.java [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package Trade;
+
+public class TradePresenter implements TradeOutputBoundary{
```
Comment (celinayueh) [inline src/main/java/Trade/TradePresenter.java:3]: Use create() to update the viewmodel. You can find the classes under the ViewModel package.

File: src/main/java/Trade/TradeInteractor.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import java.security.InvalidParameterException;
+import java.util.ArrayList;
+import java.util.List;
+/**
+ * TradeInteractor is a class that handles the trade between two players.
+ */
+public class
+TradeInteractor implements TradeInputBoundary {
+    final TradeOutputBoundary output;
+    final Campaign campaign;
```
Comment (celinayueh) [inline src/main/java/Trade/TradeInteractor.java:17]: Use CampaignAccess instead of Campaign.

**Review by sinemsoylemezoglu — COMMENTED** [review]
In some of the classes, as Celina mentioned in her comments, you also need to add the create() method, further as mentioned in the latest protocols that Max set, you can change the create() methods with performAction() method so that all of us follows the same methodology. Don't forget to make sure that you follow the latest protocols so that now, our use cases interact with each other and more importantly with CampaignAccess, instead of Campaign. Other than that good job!

File: src/main/java/tradeUseCase/tradeController.java [hunk 4]
```diff
@@ -1,24 +1,24 @@
...
 import java.util.ArrayList;
 
 
-public class TradeController {
+public class tradeController {
     //change to private`
 
-    final TradeInputBoundary input;
+    final tradeInputBoundary input;
     final Player player1;
```
Comment (celinayueh) [inline src/main/java/tradeUseCase/tradeController.java:11]: Controller should not have access to entities.

File: src/main/java/tradeUseCase/tradeInteractor.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * @param player1Properties The properties that the player who is trading is offering.
+     * @param player2Properties The properties that the player who is trading is asking for.
+     * @param player1Cash The amount of cash that the player who is trading is offering.
+     * @param player2Cash The amount of cash that the player who is trading is asking for.
+     */
+    public void trade(ArrayList<NormalProperty> player1Properties, ArrayList<NormalProperty> player2Properties,
+                      int player1Cash, int player2Cash) {
+        player1.setCash(player1.getCash() - player1Cash + player2Cash);
```
Comment (adelinewang3) [inline src/main/java/tradeUseCase/tradeInteractor.java:43]: You can also use the loseCash(amount) and gainCash(amount) method. i.e. player1.lostCash(amount), but setCash works as well

**Comment by zhenzeyang** [comment]
Good job, you also need to update the viewmodel

**Pull request closed without merging**